WireDomain/Sources/WireDomain/Synchronization/SyncManager.swift

```diff
@@ -0,0 +1,167 @@
+    private(set) var syncState: SyncState = .suspended
+    private var isSuspending = false
+
+    private let updateEventsRepository: any UpdateEventsRepositoryProtocol
+    private let updateEventProcessor: any UpdateEventProcessorProtocol
+
+    init(
+        updateEventsRepository: any UpdateEventsRepositoryProtocol,
+        updateEventProcessor: any UpdateEventProcessorProtocol
+    ) {
+        self.updateEventsRepository = updateEventsRepository
+        self.updateEventProcessor = updateEventProcessor
+    }
+
+    func performQuickSync() async throws {
+        if case .quickSync = syncState {
+            return
+        }
+
+        WireLogger.sync.info("performing quick sync")
+
+        // Opens the push channel, but events are buffered.
+        let liveEventsStream = try await updateEventsRepository.startBufferingLiveEvents()
+
+        let quickSyncTask = Task {
+            try await updateEventsRepository.pullPendingEvents()
+            try await processStoredEvents()
+        }
+
+        do {
+            syncState = .quickSync(quickSyncTask)
+            try await quickSyncTask.value
+        } catch {
+            try await suspend()
+            throw error
+        }
+
+        let liveTask = Task {
+            do {
+                for try await envelope in liveEventsStream {
+                    WireLogger.sync.info("received live event")
+                    try Task.checkCancellation()
+                    await processLiveEvents(in: envelope)
+                }
+            } catch is CancellationError {
+                WireLogger.sync.info("live task was cancelled")
+            } catch {
+                WireLogger.sync.error("live task encountered error: \(error)")
+            }
+        }
+
+        syncState = .live(liveTask)
+    }
+
+    func suspend() async throws {
+        if case .suspended = syncState {
+            return
+        }
+
+        guard !isSuspending else {
+            return
+        }
+
+        WireLogger.sync.info("suspending")
+
+        isSuspending = true
+        await closePushChannel()
+        ongoingTask?.cancel()
+        syncState = .suspended
+        isSuspending = false
+    }
+
+    private var ongoingTask: Task<Void, Error>?  {
+        switch syncState {
+        case let .quickSync(task):
+            task
+        default:
+            nil
+        }
+    }
+
+    // MARK: - Live events
+
+    private func closePushChannel() async {
+        await updateEventsRepository.stopReceivingLiveEvents()
+    }
+
+    private func processLiveEvents(in envelope: UpdateEventEnvelope) async {
+        for event in envelope.events {
+            do {
+                try await updateEventProcessor.processEvent(event)
+            } catch {
+                WireLogger.sync.error("failed to process live event, dropping: \(error)")
+            }
+        }
+
+        if !envelope.isTransient {
+            updateEventsRepository.storeLastEventEnvelopeID(envelope.id)
+        }
```

Comment:
Should we persist after processing, or after decryption?

Reply:
It should be decrypted, then stored, then the id should be persisted. I'll do this in another PR though.